packages/components/src/components/FunctionChart.tsx

```diff
@@ -27,52 +33,86 @@ function unwrap<a, b>(x: result<a, b>): a {
     throw Error("FAILURE TO UNWRAP");
   }
 }
+export type FunctionChartSettings = {
+  start: number;
+  stop: number;
+  count: number;
+};
 
-function mapFilter<a, b>(xs: a[], f: (x: a) => b | undefined): b[] {
-  let initial: b[] = [];
-  return xs.reduce((previous, current) => {
-    let value: b | undefined = f(current);
-    if (value !== undefined) {
-      return previous.concat([value]);
-    } else {
-      return previous;
-    }
-  }, initial);
+interface FunctionChartProps {
+  fn: lambdaValue;
+  chartSettings: FunctionChartSettings;
+  environment: environment;
 }
 
-export const FunctionChart: React.FC<{
-  distPlusFn: distPlusFn;
-  diagramStart: number;
-  diagramStop: number;
-  diagramCount: number;
-}> = ({ distPlusFn, diagramStart, diagramStop, diagramCount }) => {
+export const FunctionChart: React.FC<FunctionChartProps> = ({
+  fn,
+  chartSettings,
+  environment,
+}: FunctionChartProps) => {
   let [mouseOverlay, setMouseOverlay] = React.useState(0);
-  function handleHover(...args) {
-    setMouseOverlay(args[1]);
+  function handleHover(_name: string, value: unknown) {
+    setMouseOverlay(value as number);
   }
   function handleOut() {
     setMouseOverlay(NaN);
   }
   const signalListeners = { mousemove: handleHover, mouseout: handleOut };
-  let mouseItem = distPlusFn(mouseOverlay);
+  let mouseItem = runForeign(fn, [mouseOverlay], environment);
   let showChart =
-    mouseItem.tag === "Ok" ? (
+    mouseItem.tag === "Ok" && mouseItem.value.tag == "distribution" ? (
       <DistributionChart
-        distribution={mouseItem.value}
+        distribution={mouseItem.value.value}
         width={400}
         height={140}
         showSummary={false}
       />
     ) : (
       <></>
     );
-  let data1 = _rangeByCount(diagramStart, diagramStop, diagramCount);
-  let valueData = mapFilter(data1, (x) => {
-    let result = distPlusFn(x);
+  let data1 = _rangeByCount(
+    chartSettings.start,
+    chartSettings.stop,
+    chartSettings.count
+  );
+  type point = { x: number; value: result<Distribution, string> };
+  let valueData: point[] = data1.map((x) => {
+    let result = runForeign(fn, [x], environment);
     if (result.tag === "Ok") {
-      return { x: x, value: result.value };
+      if (result.value.tag == "distribution") {
+        return { x, value: { tag: "Ok", value: result.value.value } };
```

Comment:
I'm not a big fan of us having to write "value: { tag ..." everywhere. Seems like a pattern that hypothetically could be encapsulated. Not for this PR though.